api/applications.go

```diff
@@ -0,0 +1,229 @@
+// the Microsoft Graph API. It can be configured to target alternative national cloud
+// deployments via graphURI. For details on the client configuration see
+// https://learn.microsoft.com/en-us/graph/sdks/national-clouds
+func NewMSGraphApplicationClient(graphURI string, creds azcore.TokenCredential) (*MSGraphClient, error) {
+	scopes := []string{
+		fmt.Sprintf("%s/.default", graphURI),
+	}
+
+	authProvider, err := auth.NewAzureIdentityAuthenticationProviderWithScopes(creds, scopes)
+	if err != nil {
+		return nil, err
+	}
+
+	adapter, err := msgraphsdkgo.NewGraphRequestAdapter(authProvider)
+	if err != nil {
+		return nil, err
+	}
+
+	adapter.SetBaseUrl(fmt.Sprintf("%s/v1.0", graphURI))
+	client := msgraphsdkgo.NewGraphServiceClient(adapter)
+
+	ac := &MSGraphClient{
+		client: client,
+	}
+	return ac, nil
+}
+
+func (c *MSGraphClient) GetApplication(ctx context.Context, clientID string) (Application, error) {
+	filter := fmt.Sprintf("appId eq '%s'", clientID)
+	req := applications.ApplicationsRequestBuilderGetRequestConfiguration{
+		QueryParameters: &applications.ApplicationsRequestBuilderGetQueryParameters{
+			Filter: &filter,
+		},
+	}
+
+	resp, err := c.client.Applications().Get(ctx, &req)
+	if err != nil {
+		return Application{}, err
+	}
+
+	apps := resp.GetValue()
+	if len(apps) == 0 {
+		return Application{}, fmt.Errorf("no application found")
+	}
+	if len(apps) > 1 {
+		return Application{}, fmt.Errorf("multiple applications found - double check your client_id")
+	}
+
+	app := apps[0]
+
+	return getApplicationResponse(app), nil
+}
+
+func (c *MSGraphClient) ListApplications(ctx context.Context, filter string) ([]Application, error) {
+
+	req := &applications.ApplicationsRequestBuilderGetQueryParameters{
+		Filter: &filter,
+	}
+	configuration := &applications.ApplicationsRequestBuilderGetRequestConfiguration{
+		QueryParameters: req,
+	}
+	resp, err := c.client.Applications().Get(ctx, configuration)
+	if err != nil {
+		return nil, err
+	}
+
+	var apps []Application
+	for _, app := range resp.GetValue() {
+		apps = append(apps, getApplicationResponse(app))
+	}
+
+	return apps, nil
+}
+
+// CreateApplication create a new Azure application object.
+func (c *MSGraphClient) CreateApplication(ctx context.Context, displayName string) (Application, error) {
+	requestBody := models.NewApplication()
+	requestBody.SetDisplayName(&displayName)
+
+	resp, err := c.client.Applications().Post(ctx, requestBody, nil)
+	if err != nil {
+		return Application{}, err
+	}
+
+	return getApplicationResponse(resp), nil
+}
+
+// DeleteApplication deletes an Azure application object.
+// This will in turn remove the service principal (but not the role assignments).
+func (c *MSGraphClient) DeleteApplication(ctx context.Context, applicationObjectID string, permanentlyDelete bool) error {
+	err := c.client.Applications().ByApplicationId(applicationObjectID).Delete(ctx, nil)
```

Comment:
Wondering if we should check the error immediately and return it if not nil instead of combining with a multierror for the permanent delete case? I'd imagine permanent deletion will fail 100% of the time if this fails.

Reply:
I like breaking early if we receive an error. Updated in latest!